shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1075,19 +1223,108 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
-  return [FlutterTextPosition positionWithIndex:currentIndex];
+  if ([_selectionRects count] == 0) {
+    NSUInteger currentIndex = ((FlutterTextPosition*)_selectedTextRange.start).index;
+    return [FlutterTextPosition positionWithIndex:currentIndex];
+  }
+
+  FlutterTextRange* range = [[FlutterTextRange
+      rangeWithNSRange:fml::RangeForCharactersInRange(self.text, NSMakeRange(0, self.text.length))]
+      copy];
+  return [self closestPositionToPoint:point withinRange:range];
 }
 
 - (NSArray*)selectionRectsForRange:(UITextRange*)range {
-  // TODO(cbracken) Implement.
-  return @[];
+  NSUInteger start = ((FlutterTextPosition*)range.start).index;
+  NSUInteger end = ((FlutterTextPosition*)range.end).index;
+  NSMutableArray* rects = [[NSMutableArray alloc] init];
+  for (NSUInteger i = 0; i < [_selectionRects count]; i++) {
+    if ([_selectionRects[i][4] unsignedIntegerValue] >= start &&
+        [_selectionRects[i][4] unsignedIntegerValue] <= end) {
+      float width = [_selectionRects[i][2] floatValue];
+      if (start == end) {
+        width = 0;
+      }
+      CGRect rect =
+          CGRectMake([_selectionRects[i][0] floatValue], [_selectionRects[i][1] floatValue], width,
+                     [_selectionRects[i][3] floatValue]);
+      FlutterTextSelectionRect* selectionRect = [[FlutterTextSelectionRect alloc]
+          initWithRectAndInfo:rect
+             writingDirection:UITextWritingDirectionNatural
+                containsStart:(i == 0)
+                  containsEnd:(i == fml::RangeForCharactersInRange(self.text,
+                                                                   NSMakeRange(0, self.text.length))
+                                        .length)
+                   isVertical:FALSE];
+      [rects addObject:selectionRect];
+    }
+  }
+  return rects;
 }
 
 - (UITextPosition*)closestPositionToPoint:(CGPoint)point withinRange:(UITextRange*)range {
-  // TODO(cbracken) Implement.
-  return range.start;
+  NSUInteger start = ((FlutterTextPosition*)range.start).index;
+  NSUInteger end = ((FlutterTextPosition*)range.end).index;
+
+  NSUInteger _closestIndex = 0;
+  CGRect _closestRect = CGRectZero;
+  NSUInteger _closestPosition = 0;
+  float _closestY = 0;
+  float _closestX = 0;
+  for (NSUInteger i = 0; i < [_selectionRects count]; i++) {
+    NSUInteger position = [_selectionRects[i][4] unsignedIntegerValue];
+    if (position >= start && position <= end) {
+      CGRect rect =
+          CGRectMake([_selectionRects[i][0] floatValue], [_selectionRects[i][1] floatValue],
+                     [_selectionRects[i][2] floatValue], [_selectionRects[i][3] floatValue]);
+      CGPoint pointForComparison =
+          CGPointMake(rect.origin.x, rect.origin.y + rect.size.height * 0.5);
+      float yDist = abs(pointForComparison.y - point.y);
+      float xDist = abs(pointForComparison.x - point.x);
+      if (_closestIndex == 0 ||
```

Comment:
nit: pull these boolean expressions into local variables so they are annotated with english.  Or split it into static function with a clear name.